src/state/slices/opportunitiesSlice/resolvers/osmosis/utils.ts

```diff
@@ -0,0 +1,198 @@
+import type { AssetId } from '@shapeshiftoss/caip'
+import { ASSET_REFERENCE, osmosisChainId, toAssetId } from '@shapeshiftoss/caip'
+import axios from 'axios'
+import { getConfig } from 'config'
+import { bnOrZero } from 'lib/bignumber/bignumber'
+import { logger } from 'lib/logger'
+
+type OsmosisToken = {
+  denom: string
+  amount: string
+}
+
+type OsmosisPoolAsset = {
+  token: OsmosisToken
+  weight: string
+}
+
+type OsmosisPool = {
+  '@type': string
+  name: string
+  address: string
+  id: string
+  pool_params: {
+    swap_fee: string
+    exit_fee: string
+    smooth_weight_change_params: boolean
+  }
+  future_pool_governor: string
+  total_shares: {
+    denom: string
+    amount: string
+  }
+  pool_assets: OsmosisPoolAsset[]
+  total_weight: string
+  apy: string
+  tvl: string
+}
+
+type OsmosisBasePool = Omit<OsmosisPool, 'apy' | 'tvl' | 'underlyingAssets'>
+
+type OsmosisPoolList = {
+  pools: OsmosisBasePool[]
+  pagination: {
+    next_key: string | null
+    total: string
+  }
+}
+
+type PoolHistoricalData = {
+  symbol: string
+  amount: number
+  denom: string
+  coingecko_id: string
+  liquidity: number
+  liquidity_24h_change: number
+  volume_24h: number
+  volume_24h_change: number
+  volume_7d: number
+  price: number
+  fees: string
+  main: boolean
+}
+
+type PoolHistoricalDataList = {
+  [key: string]: PoolHistoricalData[]
+}
+
+const moduleLogger = logger.child({
+  namespace: ['opportunitySlice', 'resolvers', 'osmosis', 'utils'],
+})
+
+/** Somehow, the v1beta1/pools API call returns some (~12/886) objects of a type that doesn't conform
+ * to the type defined above (missing certain fields and containing others.)
+ * This type guard is used to filter out the invalid pools records. */
+const isOsmosisBasePool = (pool: any): pool is OsmosisBasePool => {
+  return (
+    pool['@type'] !== undefined &&
+    pool.address !== undefined &&
+    pool.id !== undefined &&
+    pool.pool_params !== undefined &&
+    pool.pool_liquidity === undefined &&
+    pool.future_pool_governor !== undefined &&
+    pool.total_shares !== undefined &&
+    pool.pool_assets !== undefined &&
+    pool.total_weight !== undefined
```

Comment:
Can we use `key in object` syntax? It's a better discriminator

Reply:
Good call. I used `pool.hasOwnProperty()` instead, just to be a little bit more specific.